src/charm.py

```diff
@@ -270,6 +293,95 @@ def _on_new_vault_kv_client_attached(self, event: NewVaultKvClientAttachedEvent)
+            return
+        if not self._tls_certificates_pki_relation_created():
+            logger.debug("TLS Certificates PKI relation not created, skipping")
+            return
+        common_name = self._get_config_common_name()
+        if not common_name:
+            logger.error("Common name is not set in the charm config, skipping")
+            return
+        if not vault.is_secret_engine_enabled(path=PKI_MOUNT):
+            vault.enable_pki_engine(path=PKI_MOUNT)
+        if not vault.is_intermediate_ca_set_with_common_name(
+            mount=PKI_MOUNT, common_name=common_name
+        ):
+            csr = vault.configure_pki_intermediate_ca(mount=PKI_MOUNT, common_name=common_name)
+            self.tls_certificates_pki.request_certificate_creation(
+                certificate_signing_request=csr.encode(),
+                is_ca=True,
+            )
+            self._set_pki_csr_secret_in_peer_relation(csr)
+
+    def _add_ca_certificate_to_pki_secrets_engine(self) -> None:
+        """Add the CA certificate to the PKI secrets engine."""
+        if not self.unit.is_leader():
+            logger.debug("Only leader unit can handle a vault-pki certificate request")
+            return
+        if not self._is_peer_relation_created():
+            logger.debug("Peer relation not created")
+            return
+        vault = self._get_initialized_vault_client()
+        if not vault:
+            logger.debug("Failed to get initialized Vault")
+            return
+        common_name = self._get_config_common_name()
+        if not common_name:
+            logger.error("Common name is not set in the charm config")
+            return
+        certificate = self._get_pki_ca_certificate()
+        if not certificate:
+            logger.debug("No certificate available")
+            return
+        if not vault.is_intermediate_ca_set(mount=PKI_MOUNT, certificate=certificate):
+            vault.set_pki_intermediate_ca_certificate(certificate=certificate, mount=PKI_MOUNT)
+        if not vault.is_pki_role_set(role=PKI_ROLE, mount=PKI_MOUNT):
+            vault.set_pki_charm_role(
+                allowed_domains=common_name,
+                mount=PKI_MOUNT,
+                role=PKI_ROLE,
+            )
+
+    def _get_pki_ca_certificate(self) -> str:
+        """Return the PKI CA certificate provided by the TLS provider.
+
+        Validate that the CSR matches the one in the peer relation.
+        """
+        assigned_certificates = self.tls_certificates_pki.get_assigned_certificates()
+        if not assigned_certificates:
+            return ""
+        if not self._pki_csr_secret_set_in_peer_relation():
+            logger.info("PKI CSR not set in the peer relation")
+            return ""
+        pki_csr = self._get_pki_csr_secret_in_peer_relation()
+        if not pki_csr:
+            logger.warning("PKI CSR not found in the peer relation")
+            return ""
```

Comment:
Done. Since I've started using Go, I think I now prefer returning always the same type. I've got no argument to defend this in Python though, just personal preference I think. I changed this to None here.